File: part_2-regression/multiple-linear-regression/multiple_linear_regression.py
#   Multiple Linear Regression
# To set the working directory save this .py file where we have the Data.csv file 
# and then press the Run button.  This will automatically set the working directory.
#   Importing the data from preprocessing data
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 

dataset = pd.read_csv('50_Startups.csv')

# iloc integer location based [rows, columns] : means all rows :-1 all columns except last one
X = dataset.iloc[:, :-1].values

# In python indexes are started from 0 and R starts from 1
y = dataset.iloc[:, 4].values

#   Categorical Data
#   Encoding Independent Data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,3] = labelencoder_X.fit_transform(X[:,3])
onehotencoder = OneHotEncoder(categorical_features= [3])
X = onehotencoder.fit_transform(X).toarray()

#   Avoiding Dummy Variable Trap
X = X[:, 1:] 
#In the above thing it The above column will start from 1 to end.

#Splitting the dataset into Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state =0)

#   Feature Scaling
# For multi-comment line use """ This will not be executed """ 
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

#   Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

#   Predicting the Test set results
y_pred = regressor.predict(X_test)

#   Building the model using Backword Elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis = 1)
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

#   Omit the variables which have prob more than .95
X_opt = X[:, [0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

#   Omit the variables until you have P < SL
X_opt = X[:, [0,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0,3,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0,3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

# End of Backward ELimination Algorithm

# I would like to visualize the performance of R&D vs Profit scale

File: part_2-regression/multiple-linear-regression/multiple_linear_regression.r
#   Multiple Linear Regression

# 1.    Working Directory
# 2.    Data Preprocessing Template


# Importing the Dataset
dataset = read.csv('50_Startups.csv')


#   Encoding categorical data

dataset$State= factor(dataset$State,
                         levels = c('New York', 'California', 'Florida'),
                         labels = c(1,2,3))

View(dataset)

#   Splitting the dataset into Training set and Test set

#  install.packages('caTools')
require(caTools)
library(caTools)

set.seed(123)
Split = sample.split(dataset$Profit, SplitRatio = 0.8)

training_set = subset(dataset, Split == TRUE)
test_set = subset(dataset, Split == FALSE)

#   Feature Scaling

# training_set = scale(training_set)
# test_set = scale(test_set) errors come because it contain factor values but scaling to be done to numeric objects

# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

#   Fitting the Simple Linear Regression to the Training set
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State, 
               data = training_set)
summary(regressor)

regressor = lm(formula = Profit ~ ., data = training_set) # Same as above line
# We can infer from the coefficients that R.D.Spend will have high significant impact on dependent variable Profit

#   Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)


#   Building the optimal model using Backward Elimination method

regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State, 
               data = dataset) # Here we can use dataset or training_set it is just to see the P-value
summary(regressor)

# So we see state2 and state3 will have very high p-values and can be eliminated
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend, 
               data = training_set)
summary(regressor)

# So we see Administration will have very high p-values and can be eliminated
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend, 
               data = training_set)
summary(regressor)

# So we see Marketing.Spend will have very high p-values and can be eliminated
regressor = lm(formula = Profit ~ R.D.Spend, 
               data = training_set)
summary(regressor)



